go/stellar/stellarsvc/service.go

```diff
@@ -595,10 +599,309 @@ func (s *Server) BatchLocal(ctx context.Context, arg stellar1.BatchLocalArg) (re
+			return nil, nil, err
+		}
+		local.Summary.Operations = make([]string, len(tx.Operations))
+		for i, op := range tx.Operations {
+			local.Summary.Operations[i] = stellarnet.OpSummary(op)
+		}
+	}
+
+	return &local, validated.TxEnv, nil
+}
+
+func (s *Server) ApproveTxURILocal(ctx context.Context, arg stellar1.ApproveTxURILocalArg) (txID stellar1.TransactionID, err error) {
+	mctx, fin, err := s.Preamble(ctx, preambleArg{
+		RPCName: "ApproveTxURILocal",
+		Err:     &err,
+	})
+	defer fin()
+	if err != nil {
+		return "", err
+	}
+
+	// revalidate the URI
+	vp, txEnv, err := s.validateStellarURI(mctx, arg.InputURI, http.DefaultClient)
+	if err != nil {
+		return "", err
+	}
+
+	if txEnv == nil {
+		return "", errors.New("no tx envelope in URI")
+	}
+
+	if vp.Summary.Source == "" {
+		// need to fill in SourceAccount
+		accountID, err := stellar.GetOwnPrimaryAccountID(mctx)
+		if err != nil {
+			return "", err
+		}
+		address, err := stellarnet.NewAddressStr(accountID.String())
+		if err != nil {
+			return "", err
+		}
+		txEnv.Tx.SourceAccount, err = address.AccountID()
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if txEnv.Tx.SeqNum == 0 {
+		// need to fill in SeqNum
+		sp, unlock := stellar.NewSeqnoProvider(mctx, s.walletState)
+		defer unlock()
+
+		txEnv.Tx.SeqNum, err = sp.SequenceForAccount(txEnv.Tx.SourceAccount.Address())
+		if err != nil {
+			return "", err
+		}
+	}
+
+	// sign it
+	_, seed, err := stellar.LookupSenderSeed(mctx)
+	if err != nil {
+		return "", err
+	}
+	sig, err := stellarnet.SignEnvelope(seed, *txEnv)
+	if err != nil {
+		return "", err
+	}
+
+	if vp.CallbackURL == "" {
+		_, err := stellarnet.Submit(sig.Signed)
+		if err != nil {
+			return "", err
+		}
+	} else {
+		if err := postXDRToCallback(sig.Signed, vp.CallbackURL); err != nil {
+			return "", err
+		}
+	}
+
+	return stellar1.TransactionID(sig.TxHash), nil
+}
+
+func (s *Server) ApprovePayURILocal(ctx context.Context, arg stellar1.ApprovePayURILocalArg) (txID stellar1.TransactionID, err error) {
+	mctx, fin, err := s.Preamble(ctx, preambleArg{
+		RPCName: "ApprovePayURILocal",
+		Err:     &err,
+	})
+	defer fin()
+	if err != nil {
+		return "", err
+	}
+
+	// revalidate the URI
+	vp, _, err := s.validateStellarURI(mctx, arg.InputURI, http.DefaultClient)
+	if err != nil {
+		return "", err
+	}
+
+	if vp.Amount == "" {
+		vp.Amount = arg.Amount
+	}
+
+	if vp.CallbackURL != "" {
+		recipient, err := stellar.LookupRecipient(mctx, stellarcommon.RecipientInput(vp.Recipient), arg.FromCLI)
+		if err != nil {
+			return "", err
+		}
+		if recipient.AccountID == nil {
+			return "", errors.New("recipient lookup failed to find an account")
+		}
+		recipientAddr, err := stellarnet.NewAddressStr(recipient.AccountID.String())
+		if err != nil {
+			return "", err
+		}
+
+		_, senderSeed, err := stellar.LookupSenderSeed(mctx)
+		if err != nil {
+			return "", err
+		}
+		var memoText string
+		if vp.MemoType == "MEMO_TEXT" {
+			memoText = vp.Memo
+		}
```

Comment:
Does this silently drop non-text memos?

Reply:
No, there is an error earlier (and a ticket for us to support all types).

Reply:
Seems like an easy place to forget. What about a switch with a default return error? Or at least a comment with the ticket number

Reply:
the `handle-uri` has a comment with a ticket number, i'll add here...the ticket is in the current sprint, waiting for this to be approved :) ...

Reply:
```
    if v.Memo != "" {
                if v.MemoType == "MEMO_TEXT" {
                        ui.Printf("Public memo: %q\n\n", v.Memo)
                } else {
                        // CORE-10865 will remedy this situation...
                        ui.Printf("Request contains a public memo with an unsupported type (%s): %q\n", v.MemoType, v.Memo)
                        ui.Printf("At this time, Keybase does not support that memo type.\n")

                        // shouldn't proceed if we can't include it in the transaction.
                        return nil
                }
        }
```